File: vbtcore-cpp/src/calibrator.cpp
// vbt/calibrator.cpp — 静态中位数锁死 + CV 门禁
// ============================================
// 对应 Python 端 vbtcore/calibrator.py::StaticPlateCalibrator。

#include "vbt/calibrator.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <vector>

namespace vbt {

StaticPlateCalibrator::StaticPlateCalibrator(double real_diameter_m,
                                              std::size_t min_static_frames,
                                              double max_cv)
    : real_diameter_m_(real_diameter_m),
      min_static_frames_(min_static_frames),
      max_cv_(max_cv),
      mpp_locked_(0.0),
      locked_(false) {}

void StaticPlateCalibrator::add_sample(double bbox_height_px) {
    if (!locked_ && bbox_height_px > 10.0) {
        samples_.push_back(bbox_height_px);
    }
}

bool StaticPlateCalibrator::is_ready() const {
    return samples_.size() >= min_static_frames_;
}

double StaticPlateCalibrator::lock_scale() {
    if (locked_) {
        return mpp_locked_;
    }
    if (samples_.size() < min_static_frames_) {
        throw std::runtime_error("标定样本不足");
    }

    // IQR 过滤离群点
    std::vector<double> sorted = samples_;
    std::sort(sorted.begin(), sorted.end());
    auto n = sorted.size();
    double q25 = sorted[n / 4];
    double q75 = sorted[(3 * n) / 4];
    double iqr = q75 - q25;
    std::vector<double> valid;
    for (double h : sorted) {
        if (h >= q25 - 1.5 * iqr && h <= q75 + 1.5 * iqr) {
            valid.push_back(h);
        }
    }
    if (valid.empty()) {
        throw std::runtime_error("IQR 过滤后无有效样本");
    }

    // 计算 CV（用于诊断）
    double sum = 0.0;
    for (double h : valid) sum += h;
    const double mean = sum / valid.size();
    double var_sum = 0.0;
    for (double h : valid) {
        var_sum += (h - mean) * (h - mean);
    }
    const double std = std::sqrt(var_sum / valid.size());
    const double cv = std / mean;
    (void)cv;  // CV 超限不抛错（与 Python 端一致：降级使用中位数）

    // 中位数（降级路径）
    const double median = valid[valid.size() / 2];
    mpp_locked_ = real_diameter_m_ / median;
    locked_ = true;
    return mpp_locked_;
}

}  // namespace vbt

File: vbtcore/calibrator.py
# vbtcore/calibrator.py
"""
尺度标定与门禁
==============
静态中位数锁死 + CV 变异系数晃动门禁。
在起始静止期采样检测框高度，完成门禁核验后永久冻结 mpp。
"""

from __future__ import annotations

import numpy as np


class StaticPlateCalibrator:
    def __init__(
        self,
        real_diameter_m: float = 0.45,
        min_static_frames: int = 20,
        max_cv: float = 0.015,
    ):
        """
        real_diameter_m : 杠铃片标准物理外径（默认 450 mm）
        min_static_frames : 计算标定所需的最小静止有效帧数
        max_cv : 变异系数门禁阈值；超过该值判定为机位晃动或严重遮挡
        """
        self.real_diameter_m = real_diameter_m
        self.min_static_frames = min_static_frames
        self.max_cv = max_cv
        self._samples: list[float] = []
        self._mpp_locked: float | None = None

    def add_sample(self, bbox_height: float) -> None:
        """仅在杠铃静止期输入每帧检测到的边界框高度（像素）"""
        if self._mpp_locked is None and bbox_height > 10.0:
            self._samples.append(bbox_height)

    def is_ready(self) -> bool:
        return len(self._samples) >= self.min_static_frames

    def lock_scale(self) -> float:
        """
        完成门禁核验并永久冻结 mpp。
        返回 : 米/像素（mpp）
        """
        if self._mpp_locked is not None:
            return self._mpp_locked

        if len(self._samples) < self.min_static_frames:
            raise ValueError(
                f"标定样本不足：当前 {len(self._samples)} 帧，"
                f"至少需要 {self.min_static_frames} 帧。"
            )

        heights = np.array(self._samples)

        # ── IQR 剔除离群点 ──────────────────────────────
        q25, q75 = np.percentile(heights, [25, 75])
        iqr = q75 - q25
        valid = heights[(heights >= q25 - 1.5 * iqr) & (heights <= q75 + 1.5 * iqr)]

        if valid.size == 0:
            raise RuntimeError("IQR 过滤后无有效样本，拒绝标定。")
        mean_h = float(np.mean(valid))  # noqa: pi-lens=unsafe-call
        std_h = float(np.std(valid))  # noqa: pi-lens=unsafe-call
        cv = std_h / mean_h

        # CV 超门限 → 记录警告，但使用中位数（降级运行，不直接报错）
        if cv > self.max_cv:
            import warnings

            warnings.warn(
                f"CV={cv:.4f} > {self.max_cv}（起始期有动作/晃动），"
                f"降级使用中位数标定。"
            )

        median_h = float(np.median(valid))  # noqa: pi-lens=unsafe-call
        self._mpp_locked = self.real_diameter_m / median_h
        return self._mpp_locked

    @property
    def mpp_locked(self) -> float | None:
        return self._mpp_locked
